docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
+  indirect:
+  - command: "./gradlew jib classes"
+    inputs:
+      - src/main/java
+      - src/main/resources
+    files:
+      - src: "target/classes"
+        dest: "/app/classes",
+      - src: "target/resources.
+        dest: "/app/resources",
+  - command: "compile_c_stuff.sh"
+    inputs:
+    - asdf.c
+    files:
+     - src: "asdf.o"
+       dest: "/libs/asdf.o"
+  direct:
+    - src: "local/computer/extrafiles"
+      dest: "/extrafiles"
+  includeTools: true
+}
+```
+
+Lets breakdown the different parts
+
+1. `sync: auto` - just a name for the mode, can be anything
+2. `indirect` - this means, you can't sync files directly, you need a command to
+   tranform them before sync. Indirect is also a list that can handle multiple
+   indirect types.
+     1. `command` - the command to run to generate syncable files
+     2. `inputs` - the inputs to watch to trigger `command`, if left empty, this
+        should just be considered the full list of inputs that are normally
+        watched for this build minus anything in `direct` (below).
+     3. `files` - the built files to sync (and where to sync them)
+3. `direct` - same as manual sync, just copy them directly, don't wait on some
```

Comment:
Would this replace manual? or another definition in `auto` ?

Currently, we don't support `manual` and `infer` sync rules together. 
https://skaffold.dev/docs/how-tos/filesync/#limitations
As per this proposal, do we plan to support `auto` as well `manual` sync modes?

Reply:
Given that sync does not currently support multiple sync configurations, this is designed is that this is all one large block that does many things. A builder can have some files that need to be compiled, and others that don't -- so if I was configuring this from jib, I could separate them out.

However this design may be flawed in that it redoes the work that other sync options already do.